### PR TITLE
Rewrite pattern matching to use explicit length

### PR DESCRIPTION
**Pull request by ColtonWilley**
# Description
Rewrite pattern matching to always use explicit lengths instead of expecting NULL terminated strings, thus replicating the behavior of openssl X509_check_host(). This rewrite fixes a buffer overread of size 1 to read the NULL terminator, in addition to fixing the issue of some strings matching or not matching incorrectly.

Fixes ZD 17987 which was only partially fixed by PR#7534.

# Checklist

 - [X] added tests

### CODE REVIEW

**Review by dgarske — CHANGES_REQUESTED** [review]
Really nice work!

**Comment by ColtonWilley** [comment]
Retest this please

**Review by dgarske — APPROVED** [review]

File: src/x509.c [hunk 1]
```diff
@@ -13384,6 +13385,22 @@ int wolfSSL_X509_check_host(WOLFSSL_X509 *x, const char *chk, size_t chklen,
         goto out;
     }
 
+    /* Replicate openssl behavior for checklen */
+    if (chklen == 0) {
+        chklen = (size_t)(XSTRLEN(chk));
+    }
+    else {
+        for (i = 0; i < (chklen > 1 ? chklen - 1 : chklen); i++) {
```
Comment (dgarske) [inline src/x509.c:13393]: Consider making `i` a `size_t` to match `chklen`.

File: src/internal.c [hunk 2]
```diff
@@ -15410,9 +15437,12 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
...
                     else {
                         if (MatchDomainName(
-                                 args->dCert->subjectCN,
-                                 args->dCert->subjectCNLen,
-                                 (char*)ssl->buffers.domainName.buffer) == 0) {
+                                args->dCert->subjectCN,
+                                args->dCert->subjectCNLen,
+                                (char*)ssl->buffers.domainName.buffer,
+                                (ssl->buffers.domainName.buffer == NULL ? 0 :
+                                (word32)XSTRLEN((const char *)ssl->buffers.domainName.buffer))) == 0)
```
Comment (dgarske) [inline src/internal.c:15444]: Check line length. Max 80 chars.

**Review by dgarske — CHANGES_REQUESTED** [review]
Again, great work on this PR. Nit picks only

**Review by dgarske — APPROVED** [review]

**Pull request merged**